### ISSUE TITLE
Netflix and Chill: Building a Recommendation System in Excel

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9329]
Netflix and Chill: Building a Recommendation System in Excel<br>
<br>
https://ift.tt/rI8aoUG<br>
<br>
Dave Smith<br>
<br>

<div><div><div><div><div>
<h1>1.1 The Data</h1>
<p>The blog post model uses 30 fictitious ratings (5 users x 6 movies) to simplify the tutorial.</p>
<ul><li>To follow along and experiment with the model as we go, you can download the spreadsheets (Excel or Google Sheets) <a rel="noopener ugc nofollow" href="https://drive.google.com/open?id=1y4X8H56TS6M7AXAU7yIm0EyxhqNUy1sz" target="_blank">here</a>.</li></ul>
<p>The Fast.AI lesson uses 225 real ratings (which is tough to illustrate in a blog post) based on the 15 most watched movies and the 15 most active users. With the exception of fake ‘user names’, the ratings <a rel="noopener ugc nofollow" href="http://files.grouplens.org/datasets/movielens/ml-latest-small.zip" target="_blank">data is from MovieLens</a>.</p>
<ul>
<li>To see the model (Excel or Google Sheets) based on the Fast.AI lesson, you can download that <a rel="noopener ugc nofollow" href="https://drive.google.com/open?id=1aY7ebZ2xb0UGZQ_-iiR7VnEECleZXd8J" target="_blank">here</a>.</li>
<li><em>Note — The Fast.AI model uses 15 epochs instead of 50 (due to more data) and the calculations work fairly slow in Google Sheets so Excel is recommended.</em></li>
</ul>
<h1>1.2 Splitting the Data — Training vs. Test</h1>
<p>We’ll use 25 ratings to train our model and 5 ratings to test the model’s accuracy.</p>
<p>Our goal is to build a system that works well on the 25 known ratings (training data) and <strong>hope </strong>it predicts well on the 5 hidden (but known) ratings (test data).</p>
<p>If we had more data, we’d split our data into 3 groups — training (~70%), validation (~20%), and test (~10%) and we would use our validation set to validate our model.</p>
<h1><strong>1.3 Rating Prediction Formula</strong></h1>
<p>The rating prediction is a matrix multiplication or “dot product” of the user/movie features + user bias + movie bias.</p>
<figure></figure><p>Formal definition:</p>
<figure></figure><h2><strong>1.3.1 User/movie features</strong></h2>
<ul>
<li>Intuitively, these features represent things like genre, actors, length of movie, director, year of movie, etc. Even though we don’t quite know what each feature is, we can intuitively guess what they could represent after we visualize them on a graph (see PART 4).</li>
<li>I used 3 features for simplicity, but an actual model could have 50, 100, or more. The # of features will grow in proportion to the complexity of your data. With too many features, the model will “overfit/memorize” your training data and it won’t generalize well when it predicts the hidden (but known) test ratings.</li>
<li>If a user had a high value for their 1st feature (let’s assume it represents “comedies”) and the the movie also had a high value for this “comedy” feature, this would likely result in a high rating for that movie.</li>
</ul>
<h2><strong>1.3.2 User/movie biases</strong></h2>
<p>A user bias is how harsh or nice the movie critic is. If the average rating of all movies on Netflix is a 3.5 and the average of everything you rate is a 4.0, you would have a 0.5 bias. Movie bias can be thought of the same way. If the <em>Titanic </em>had an average rating of 4.25 across all users, it would have a 0.75 bias (= 4.25–3.50).</p>
<h1><strong>1.4 RMSE — Evaluating Prediction Accuracy</strong></h1>
<p>RMSE = Root Mean Squared Error</p>
<p>RMSE is a number that attempts to answer the question “on average, how many stars (1–5) did your predicted ratings differ vs. the actual ratings”?</p>
<p>The lower the RMSE, the better the predictions…</p>
<figure></figure><p><strong>Observations:</strong></p>
<ul>
<li>
<strong>We only care about the absolute differences</strong>. A prediction that is 1 higher than the actual rating has the same error, 1, as a prediction that is 1 lower than the actual rating.</li>
<li>
<strong>RMSE is an average of <em>magnitude of the error </em>which isn’t the same as the <em>absolute average error</em>.</strong> In our example above, the absolute average error was 0.75 (1 + 1 + 0.25 = 2.25/3 = 0.75), but the RMSE was 0.8292. RMSE gives a higher weight to large errors which is useful when large errors are undesirable.</li>
</ul>
<p>Formal RMSE definition:</p>
<figure></figure><h1>1.5 Model Inputs (Hyperparameter Tuning)</h1>
<p>In the spreadsheet, there are drop-down filters for 3 inputs. You should play around with each of these inputs and look at the impact they have on your errors.</p>
<ul>
<li>
<strong>Training epoch </strong>— 1 epoch is 1 training loop through the the entire data set</li>
<li>
<strong>Learning rate </strong>— Controls how fast we adjust the weights/biases</li>
<li>
<strong>L2 (lambda) penalty factor </strong>— A term to help the model prevent overfitting or “memorizing” the training data so it can generalize on unseen test data</li>
</ul>
<figure></figure><p>Now, let’s watch the <em>magic </em>as the model starts off with random weights and learns the optimal ones.</p>
</div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Medium <br>
<br>
May 15, 2023 at 03:59PM
